source/reference/data-types.txt

```diff
@@ -38,325 +40,451 @@ string or as a ``Date`` object:
-```````````````````````
 
-To return the date as a string, use the ``Date()`` method, as in the
-following example:
+ObjectId
+--------
+
+``mongosh`` provides the ``ObjectId()`` wrapper class around the
+:ref:`objectid` data type. To generate a new ObjectId, use the following
+operation in ``mongosh``:
 
 .. code-block:: javascript
 
-   var myDateString = Date();
+   new ObjectId
 
-To print the value of the variable, type the variable name in the
-shell, as in the following:
+.. seealso::
 
-.. code-block:: javascript
+   :method:`ObjectId` 
 
-   myDateString
-
-The result is the value of ``myDateString``:
+.. _shell-type-int:
 
-.. code-block:: javascript
+Int32
+-----
 
-   Wed Dec 19 2012 01:03:25 GMT-0500 (EST)
+If a number can be converted to a 32-bit integer, ``mongosh`` will
+store it as ``Int32``. If not, ``mongosh`` defaults to storing the
+number as a ``Double``. Numerical values that are stored as ``Int32``
+in ``mongosh`` would have been stored by default as ``Double`` in the
+``mongo`` shell.
 
-To verify the type, use the ``typeof`` operator, as in the following:
+The :node-api:`Int32() <Int32.html>` constructor can be used to
+explicitly specify 32-bit integers.
 
 .. code-block:: javascript
 
-   typeof myDateString
+      db.types.insertOne(
+                          {
+                              "_id": 1,
+                              "value": Int32("1"),
+                              "expectedType": "Int32" 
+                          }
+                        )
 
-The operation returns ``string``.
+.. warning::
 
-Return ``Date``
-```````````````
+   Default ``Int32`` and ``Double`` types may be stored inconsistently
+   if you connect to the same collection using both ``mongosh`` and the
+   legacy ``mongo`` shell. 
 
-``mongosh`` wraps objects of ``Date`` type with the
-``ISODate`` helper; however, the objects remain of type ``Date``.
+.. _shell-type-long:
 
-The following example uses both the ``new Date()`` constructor and the
-``ISODate()`` constructor to return ``Date`` objects.
+Long
+----
 
-.. code-block:: javascript
+The :node-api:`Long() <Long.html>` constructor can be used to
+explicitly specify a 64-bit integer.
 
-   var myDate = new Date();
-   var myDateInitUsingISODateWrapper = ISODate();
+.. code-block:: javascript
 
-You can use the ``new`` operator with the ``ISODate()`` constructor as
-well.
+      db.types.insertOne(
+                          {
+                              "_id": 3,
+                              "value": Long("1"),
+                              "expectedType": "Long" 
+                          }
+                        )
 
-To print the value of the variable, type the variable name in the
-shell, as in the following:
+.. note::
 
-.. code-block:: javascript
+   In the legacy :binary:`~bin.mongo` shell ``NumberLong()`` accepted
+   either a string or integer value. In ``mongosh``, ``NumberLong()``
+   only accepts string values. :node-api:`Long() <Long.html>` provides
```

Comment:
I'm a little confused by the inclusion of this last sentence about `Long`. It seems to come a little out of the blue since we were just talking about a different type (`NumberLong()`). Maybe I'm misunderstanding something here though, so maybe you can just clarify in a comment if you and @addaleax feel this is clear enough as written.

Reply:
`Long(string)` is the preferred version to create an int64 value – `NumberLong()` is the “legacy” version that is also provided by the old shell.

Reply:
Sounds good. Thanks for the clarifications!